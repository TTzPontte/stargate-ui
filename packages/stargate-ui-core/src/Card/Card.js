import React, { useState } from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';

import Factory from '../Factory';

const styles = (theme) => {
   /**
   * @todo improve dynamic properties
   */
  const {
    active,
    spacing,
    palette,
    border,
    radius,
    mode,
    resets,
  } = theme;
  const { setLightness } = palette;

  const getColor = (color, type = 'color') => (
    palette?.[color]?.[mode][type]
  );

  const getTextColor = (color) => (
    color && palette[color][mode].text
  );

  const card = {
    display: 'flex',
    maxWidth: 'max-content',
    border: ({borderColor, color}) => {
      const colorOfBorder = borderColor || color;
      return [
        [
          1,
          'solid',
          getColor(colorOfBorder),
        ],
      ]
    },
    cursor: ({ clickable }) => (
      clickable && 'pointer'
    ),
    borderRadius: radius(2),
    '&:not($cardSelected)': {
      backgroundColor: ({ color }) => (
        color === 'default' ? palette.colors.grey[100] : getColor(color)
        ),
      },
      color: ({ color }) => (
        getTextColor(color)
        ),
  };

  const cardSelected = {
    backgroundColor: 'red',
  };

  return {
    cardSelected,
    card,
  };
};

const Card = (props) => {
  const {
    selected: defaultValue = false,
    color = 'default',
    borderColor = 'default',
    onChange = () => {},
    clickable,
    ...factoryProps
  } = props;

  const cardProps = {};

  const [{
    card,
    cardSelected,
  }] = useStyles(styles, {
    color,
    borderColor,
    clickable,
  });

  const [selected, setSelected] = useState(defaultValue);

  const handleClick = () => {
    setSelected(!selected);
    onChange({ selected });
  }

  if (clickable) {
    cardProps.onClick = handleClick;
  }

  const classCard = clsx({
    [cardSelected]: clickable && selected,
  }, card)

  return (
    <Factory
      paddingX={2}
      paddingY={4}
      {...factoryProps}
      {...cardProps}
      className={classCard}
    />
  );
};

export default Card;
