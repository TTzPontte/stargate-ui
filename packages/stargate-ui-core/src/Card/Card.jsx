import React, { useState } from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';
import PropTypes from 'prop-types';

import Factory from '../Factory';

const styles = (theme) => {
  const {
    palette,
    radius,
    mode,
  } = theme;
  const getColor = (color, type = 'color') => (
    palette?.[color]?.[mode][type]
  );

  const getVariant = (variant, shade) => (
    palette?.colors[variant]?.[shade]
  );

  const getTextColor = (color) => (
    color && palette[color][mode].text
  );

  const card = {
    display: 'inline-block',
    cursor: ({ clickable }) => (
      clickable && 'pointer'
    ),
    borderRadius: ({ borderRadius }) => (
      borderRadius ? radius(borderRadius) : radius(2)
    ),
    '&:not($cardSelected)': {
      backgroundColor: ({...props}) => {
        const { color, variant, shade } = props;
        if (variant) {
          return getVariant(variant, shade);
        }
        return color === 'default' ? palette.colors.grey[100] : getColor(color)
      },
      color: ({ color, variant, shade }) => {
        if (variant && variant !== 'grey' && shade >= 700) {
          return palette.colors.grey[100]
        }
        return getTextColor(color)
      },
      border: ({ borderColor, color }) => {
        const colorOfBorder = borderColor || color;
        return [
          [
            1,
            'solid',
            getColor(colorOfBorder),
          ],
        ];
      },
    },
  };

  const cardSelected = {
    backgroundColor: getColor('primary'),
    borderColor: getColor('primary'),
    color: getTextColor('primary'),
  };

  return {
    card,
    cardSelected,
  };
};

const Card = (props) => {
  const {
    selected: defaultValue = false,
    color = 'default',
    variant,
    shade = 800,
    borderColor,
    borderRadius,
    onChange = () => {},
    clickable,
    className: inheritedClassName,
    ...factoryProps
  } = props;

  const cardProps = {};

  const [{
    card,
    cardSelected,
  }] = useStyles(styles, {
    color,
    shade,
    variant,
    clickable,
    borderColor,
    borderRadius,
  });

  const [selected, setSelected] = useState(defaultValue);

  const handleClick = () => {
    setSelected(!selected);
    onChange({ selected });
  }

  if (clickable) {
    cardProps.onClick = handleClick;
  }

  const classCard = clsx(card, {
    [cardSelected]: clickable && selected,
  },
    inheritedClassName,
  );

  return (
    <Factory
      {...factoryProps}
      {...cardProps}
      className={classCard}
    />
  );
};

Card.displayName = 'Card';

Card.propTypes = {
  /**
   * Add custom borderColor style.
   * **@default** `undefined`
   */
  borderColor: PropTypes.string,
  /**
   * Add custom variant of color style.
   * **@default** `undefined`
   */
  variant: PropTypes.string,
  /**
   * Set shade of variant color style.
   * **@default** `800`
   */
  shade: PropTypes.number,
  /**
   * Add custom borderRadius to style card.
   * **@default** `undefined`
   */
  borderRadius: PropTypes.number,
  /**
   * Enable card to be clicked as a button.
   * **@default** `undefined`
   */
  clickable: PropTypes.bool,
  /**
   * Add selected style.
   * **@default** `false`
   */
  selected: PropTypes.bool,
  /**
   * Add color style to background and border.
   * **@default** `default`
   */
  color: PropTypes.oneOf([
    'default',
    'primary',
    'secondary',
    'success',
    'warning',
    'info',
    'error',
    false,
  ]),
  /**
   * Trigger when element is changed.
   * **@default** `func`
   */
  onChange: PropTypes.func,
};

export default Card;
