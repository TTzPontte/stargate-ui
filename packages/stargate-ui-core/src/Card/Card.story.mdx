import {
  Meta,
  Canvas,
  Status,
  Story,
  withAction,
  Props,
} from '@pontte/stargate-ui-playground';

import Factory from '../Factory';
import Typography from '../Typography';
import Card from './Card';

<Meta title="Core/Card" component={Card} />

# Card

<Status type="stable" />

## Basic Usage

<Canvas>
  <Story name="Basic Usage" parameters={{
    jest: ['Card.spec.jsx'],
  }}>
    {
      () => (
        <Card
          onClick={withAction('onClick')}
          onChange={withAction('onChange')}
          paddingX={8}
          paddingY={6}
          clickable
        >
          <Typography element="span">
            Click
          </Typography>
        </Card>
      )
    }
  </Story>
</Canvas>

<Props component={Card} />

### Default Card

<Canvas>
  <Card
    paddingX={4}
    paddingY={6}
    variant="grey"
    shade={800}
    borderColor="transparent"
  >
    <Typography element="span">
      Read Only
    </Typography>
  </Card>
  <Card
    paddingX={4}
    paddingY={6}
    clickable
    variant="pink"
  >
    <Typography element="span">
     Click
    </Typography>
  </Card>
</Canvas>

### Customized Card

<Canvas>
  <Card
    paddingX={4}
    paddingY={2}
    color="warning"
    borderColor="success"
    clickable
    borderRadius={4}
  >
    <Typography element="span">
      Click
    </Typography>
  </Card>
  <Card
    paddingX={2}
    paddingY={2}
    clickable
    selected
  >
    <Factory>
      <Typography
        element="span"
        variant="h4"
        display="block"
        textAlign="center"
      >
        180 meses
      </Typography>
      <Typography
        element="span"
        variant="small"
        display="block"
        gutter={.5}
        textAlign="center"
      >
        1ª PARCELA
      </Typography>
      <Typography
        element="span"
        variant="h4"
        display="block"
        textAlign="center"
      >
        R$ 936,62
      </Typography>
      <Typography
        element="span"
        variant="small"
        display="block"
        gutter={.5}
        textAlign="center"
      >
        ÚLTIMA PARCELA
      </Typography>
      <Typography
        element="span"
        variant="h4"
        display="block"
        textAlign="center"
      >
        R$ 336,62
      </Typography>
    </Factory>
  </Card>
</Canvas>
