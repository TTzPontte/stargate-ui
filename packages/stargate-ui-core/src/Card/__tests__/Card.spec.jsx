import React from 'react';

import Card from '../Card';

describe('Card', () => {
  it('should render properly', () => {
    const { container } = render(
      <Card clickable>
        Click
      </Card>,
    );

    expect(container).toMatchSnapshot();
  });
});
