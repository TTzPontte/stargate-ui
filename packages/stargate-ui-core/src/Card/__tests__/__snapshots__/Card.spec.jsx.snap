// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render properly 1`] = `
<div>
  <div
    class="stargate-ui-factory-0-2-4 stargate-ui-factory-d0-0-2-6 stargate-ui-card-0-2-1 stargate-ui-card-d0-0-2-3"
  >
    Click
  </div>
</div>
`;
